# Analyzing Default Risk: Case Study of Islamic Bank <a id='intro'></a>

A `Islamic Bank` that implements `religion of Islam` shariah contract in its business processes where it only provides buying and selling services, not a contract of borrowing money with usury to the first party (`developer / seller`), second (`bank`) and third (` customer`). Customers who want something but are constrained by costs can use this facility, where goods are purchased by the bank and then resold at a higher margin to the customer by paying installments over a certain period of time. Our job is to prepare reports for the business division of a bank. We will find out the effect of a customer's marital status and the number of children he has on the probability of default in repayment of installments. The bank already has some data regarding the eligibility of the customer. This data was not real from Islamic Bank and was only for learning purposes.

The main objective is to find out how big the risk of 'Default Paying' is carried out by customers based on several categories including 'Income Level', 'Family Status', 'Installment Purposes' and 'Number of Children' to be used by companies to assess potential customers. The hypotheses made for the `default` analysis include:
- Is it true that the highest rating is `Default` by customers with `Low Income Level`?
- Is it true that the highest rating is `Default` by customers with family status `Unmarried`?
- Is it true that the highest `Default` rating by customers with `Estate Purposes` installment?
- Is it true that the highest rating of `Default` by customers is with customers who are `No Children`?

# Content <a id='back'></a>

* [Intro](#intro)
* [Stage 1. Data Exploration](#cont_1)
     * [1.1 Open *data file* and data general information](#cont_2)
     * [1.2 Interim conclusion](#cont_3)
* [Stage 2. Data Transformation](#cont_4)
     * [2.1 Working with missing values](#cont_5)
     * [2.2 Fixed missing values in **{total_income}**](#cont_6)
     * [2.3 Fixed value in **{days_employed}**](#cont_7)
     * [2.4 Data Categorization](#cont_8)
* [Stage 3. Checking the Hypothesis](#cont_9)
     * [3.1 Hypothesis 1: Is there a correlation between having children and the probability of defaulting?](#cont_10)
     * [3.2 Hypothesis 2: Is there a correlation between family status and the probability of defaulting?](#cont_11)
     * [3.3 Hypothesis 3: Is there a correlation between income levels and the probability of defaulting?](#cont_12)
     * [3.4 Hypothesis 3: How do installment objectives affect the default rate?](#cont_13)
* [Stage 4. General Conclusion](#cont_14)

## Open data *file* and general data information <a id= 'cont_2' ></a>

The libraries used are `pandas` as dataframe analysis and `numpy` as the generator of nan values. The dataset used comes from `customers_scoring_eng.csv`. This dataset is not an original dataset from Islamic banks and it is only an analytical exercise.

In [1]:
# load the required libraries
import pandas as pd
import numpy as np

In [2]:
# load datasets
df = pd.read_csv('customers_scoring_eng.csv')

## Data Exploration <a id= 'cont_1' ></a>

**Data Description**
The dataset contains various variables/columns which are detailed as follows:
- `children` - the number of children in the family
- `days_employed` - the customer's work experience in days
- `dob_years` - customer's age in years
- `education` - level of customer education
- `education_id` - identifier for the customer's education level
- `family_status` - an identifier for the customer's marital status
- `family_status_id` - marital status identifier
- `gender` - the customer's gender
- `income_type` - a type of work
- `debt` - has the customer ever defaulted on a loan?
- `total_income` - monthly income
- `purpose` - the purpose of getting a loan

In [3]:
# displays dataset information
df.info()

# displays sample datasets
df. head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id   
0         1   -8437.673028         42    bachelor's degree             0  \
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income   
0            married                 0      F    employee     0     40620.102  \
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

The number of rows and columns in the `credit_score` dataframe is `21525 rows` and `12 columns`

From the dataframe, several errors need to be corrected to make it easier to analyze the data:
- In the `days_employed` column there is a negative number and it is still a float number
- In the `education` column, each value contains the same meaning but is written differently in uppercase and lowercase letters
- In the `total_income` column, values still use a long float number

**Filtering Missing Data**

In [4]:
# let's see the filtered table with missing values in the second column containing missing data
df_filter = df[df['days_employed'].isna()]

# displays missing sample datasets
df_filter.head()

children  days_employed  dob_years            education  education_id   
12         0            NaN         65  secondary education             1  \
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   

        family_status  family_status_id gender    income_type  debt   
12  civil partnership                 1      M        retiree     0  \
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   

    total_income                   purpose  
12           NaN         to have a wedding  
26           NaN                 education  
29           NaN    building a real estate  
41           NaN  second-hand car purchase  
55           NaN         to have a wedding

In [5]:
# calculates the number of missing values
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

From the information above, it can be seen that there is an indication of similarity between `days_employed` and `total_income` for the `missing` value by checking the value of `empty` and the same amount as `2174`. Next, we will prove it.

In [6]:
# checks the missing value relation between the `days_employed` and `total_income` columns
df_filter_simmetry = df.loc[(df['days_employed'].isna()) & (df['total_income'].isna())]

# sum of missing values in the `days_employed` and `total_income` columns
print('days_employed -->', df_filter_simmetry['days_employed'].isna().sum())
print('total_income -->', df_filter_simmetry['total_income'].isna().sum())

days_employed --> 2174
total_income --> 2174


In [7]:
# calculates the percentage of missing values
mean_missing_values = df['days_employed'].isna().sum()/len(df)
print(f'Missing Values --> {mean_missing_values:.2%}')

Missing Values --> 10.10%


After checking, indeed, the sum of the `missing` values in the `days_employed` column is the same as the `lost` values in the `total_income` column of `2174` and the percentage of missing values is `10.10%`. This indicates that the `missing` value is not a random value and indicates a technical error. Let's prove it.

**Check Distribution on DF_Filter**

In [8]:
# distribution of children
child_count = df_filter["children"].value_counts()
child_percent = df_filter["children"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : child_count.index.map(child_percent), "count" : child_count}).rename_axis("unique")

percent%  count
unique                
 0       66.19%   1439
 1       21.85%    475
 2        9.38%    204
 3        1.66%     36
 20       0.41%      9
 4        0.32%      7
-1        0.14%      3
 5        0.05%      1

In [9]:
# distribution of education
edu_count = df_filter["education"].value_counts()
edu_percent = df_filter["education"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : edu_count.index.map(edu_percent), "count" : edu_count}).rename_axis("unique")

percent%  count
unique                             
secondary education   64.77%   1408
bachelor's degree     22.82%    496
SECONDARY EDUCATION    3.08%     67
Secondary Education    2.99%     65
some college           2.53%     55
Bachelor's Degree      1.15%     25
BACHELOR'S DEGREE      1.06%     23
primary education      0.87%     19
Some College           0.32%      7
SOME COLLEGE           0.32%      7
Primary Education      0.05%      1
PRIMARY EDUCATION      0.05%      1

In [10]:
# distribution of family_status_id
famid_count = df_filter["family_status_id"].value_counts()
famid_percent = df_filter["family_status_id"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : famid_count.index.map(famid_percent), "count" : famid_count}).rename_axis("unique")

percent%  count
unique                
0         56.9%   1237
1        20.33%    442
4        13.25%    288
3         5.15%    112
2         4.37%     95

In [11]:
# distribution of income_type
income_count = df_filter["income_type"].value_counts()
income_percent = df_filter["income_type"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : income_count.index.map(income_percent), "count" : income_count}).rename_axis("unique")

percent%  count
unique                       
employee        50.83%   1105
business        23.37%    508
retiree          19.0%    413
civil servant    6.76%    147
entrepreneur     0.05%      1

From the findings above, the percentage of `missing` values is `10%` from the dataset, this value is quite a large value and cannot be simply eliminated. the possible cause of the missing values can be seen from the distribution of the `education` column which is still mixed with upper and lower case values (can't tell the difference).

**Check Distribution on DF**

In [12]:
# distribution of children
child_all_count = df["children"].value_counts()
child_all_percent = df["children"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : child_all_count.index.map(child_all_percent), "count" : child_all_count}).rename_axis("unique")

percent%  count
unique                
 0       65.73%  14149
 1       22.38%   4818
 2        9.55%   2055
 3        1.53%    330
 20       0.35%     76
-1        0.22%     47
 4        0.19%     41
 5        0.04%      9

In [13]:
# distribution of education_id
edu_id_all_count = df["education_id"].value_counts()
edu_id_all_percent = df["education_id"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : edu_id_all_count.index.map(edu_id_all_percent), "count" : edu_id_all_count}).rename_axis("unique")

percent%  count
unique                
1        70.77%  15233
0        24.44%   5260
2         3.46%    744
3         1.31%    282
4         0.03%      6

In [14]:
# distribution of education
edu_all_count = df["education"].value_counts()
edu_all_percent = df["education"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : edu_all_count.index.map(edu_all_percent), "count" : edu_all_count}).rename_axis("unique")

percent%  count
unique                             
secondary education   63.88%  13750
bachelor's degree     21.92%   4718
SECONDARY EDUCATION    3.59%    772
Secondary Education     3.3%    711
some college            3.1%    668
BACHELOR'S DEGREE      1.27%    274
Bachelor's Degree      1.25%    268
primary education      1.16%    250
Some College           0.22%     47
SOME COLLEGE           0.13%     29
PRIMARY EDUCATION      0.08%     17
Primary Education      0.07%     15
graduate degree        0.02%      4
Graduate Degree         0.0%      1
GRADUATE DEGREE         0.0%      1

In [15]:
# distribution of family_status_id
fam_id_all_count = df["family_status_id"].value_counts()
fam_id_all_percent = df["family_status_id"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : fam_id_all_count.index.map(fam_id_all_percent), "count" : fam_id_all_count}).rename_axis("unique")

percent%  count
unique                
0        57.51%  12380
1        19.41%   4177
4        13.07%   2813
3         5.55%   1195
2         4.46%    960

In [16]:
# distribution of income_type
income_all_count = df["income_type"].value_counts()
income_all_percent = df["income_type"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : income_all_count.index.map(income_all_percent), "count" : income_all_count}).rename_axis("unique")

percent%  count
unique                                     
employee                      51.66%  11119
business                      23.62%   5085
retiree                       17.91%   3856
civil servant                  6.78%   1459
unemployed                     0.01%      2
entrepreneur                   0.01%      2
student                         0.0%      1
paternity / maternity leave     0.0%      1

After checking the distribution of the `original dataframe` dataset and the distribution of the `datasets filter` dataset it shows the exact similarity of the pattern that for each `datasets filter` value is `10x times` that of the original `original dataframe` datasets. There is a strong indication that the `missing` values are not random.

**Check Distribution of `family_status` Column**

In [17]:
# look for other patterns
# distribution of family_status dataframe
famstat_all_count = df["family_status"].value_counts()
famstat_all_percent = df["family_status"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : famstat_all_count.index.map(famstat_all_percent), "count" : famstat_all_count}).rename_axis("unique")

percent%  count
unique                           
married             57.51%  12380
civil partnership   19.41%   4177
unmarried           13.07%   2813
divorced             5.55%   1195
widow / widower      4.46%    960

In [18]:
# distribution of family_status dataframe_filter
famstat_count = df_filter["family_status"].value_counts()
famstat_percent = df_filter["family_status"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : famstat_count.index.map(famstat_percent), "count" : famstat_count}).rename_axis("unique")

percent%  count
unique                           
married              56.9%   1237
civil partnership   20.33%    442
unmarried           13.25%    288
divorced             5.15%    112
widow / widower      4.37%     95

**Check Distribution of `education_id` Column**

In [19]:
# distribution of education_id dataframe
eduid_all_count = df["education_id"].value_counts()
eduid_all_percent = df["education_id"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : eduid_all_count.index.map(eduid_all_percent), "count" : eduid_all_count}).rename_axis("unique")

percent%  count
unique                
1        70.77%  15233
0        24.44%   5260
2         3.46%    744
3         1.31%    282
4         0.03%      6

In [20]:
# distribution of education_id dataframe_filter
eduid_count = df_filter["education_id"].value_counts()
eduid_percent = df_filter["education_id"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : eduid_count.index.map(eduid_percent), "count" : eduid_count}).rename_axis("unique")

percent%  count
unique                
1        70.84%   1540
0        25.02%    544
2         3.17%     69
3         0.97%     21

**Temporary conclusion** <a id= 'cont_3' ></a>

From findings of the `education_id` column indicate that the pattern from the `datasets filter` does not have all the values of the `original datasets` which means that there is no special value associated with the `empty` value and is not a coincidence.

In [21]:
# distribution of income_type dataftame
inc_all_count = df['income_type'].value_counts()
inc_all_percent = df['income_type'].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : inc_all_count.index.map(inc_all_percent), "count" : inc_all_count}).rename_axis("unique")

percent%  count
unique                                     
employee                      51.66%  11119
business                      23.62%   5085
retiree                       17.91%   3856
civil servant                  6.78%   1459
unemployed                     0.01%      2
entrepreneur                   0.01%      2
student                         0.0%      1
paternity / maternity leave     0.0%      1

In [22]:
# distribution of income_type dataftame_filter
inc_all_count = df_filter['income_type'].value_counts()
inc_all_percent = df_filter['income_type'].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : inc_all_count.index.map(inc_all_percent), "count" : inc_all_count}).rename_axis("unique")

percent%  count
unique                       
employee        50.83%   1105
business        23.37%    508
retiree          19.0%    413
civil servant    6.76%    147
entrepreneur     0.05%      1

**Temporary conclusions**

The `income_type` column indicates that the `student` and `paternity / maternity leave` values still have `total_income` income compared to other values. It can be concluded that the `missing` values are due to a technical error and are not random values. To overcome this, the following steps are required:
- Since the missing values are `numeric`, it is necessary to search for statistical values such as `mean` and `median`.
- Considers which value to use to fill in the missing values such as the `mean` or `median` value.
- Duplicate checking
- Remove duplicates
- Change the data type

## Data Transformation <a id= 'cont_4' ></a>

Next, let's fix the errors in the dataset to make it easier to do the next analysis.

In [23]:
# view unique values in the `education` column
df['education'].value_counts()

education
secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: count, dtype: int64

The value in the `education` column is still not calculated correctly because the font format is not uniform. Let's change it to a uniform letter.

In [24]:
# changes all values in the `education` column to lowercase
df['education'] = df['education'].str.lower()

In [25]:
# check the `education` column to ensure that everything is fixed
df['education'].value_counts()

education
secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: count, dtype: int64

The value in the `education` column can be calculated correctly because the font format is uniform.

In [26]:
# checks `children` for suspicious values
df['children'].value_counts()

children
 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: count, dtype: int64

For the number of children in the `children` column, there is an odd value, namely `'-1'`, this may be due to a user input error who accidentally entered the `'-'` sign. Let's change it to the positive form by removing the `'-'` sign.

In [27]:
# change the value of `-1` to `1` in the children column
df.loc[df['children'] == -1, 'children'] = 1

In [28]:
# check the `children` column to ensure that everything is fixed
df['children'].unique()

array([ 1,  0,  3,  2,  4, 20,  5], dtype=int64)

The `children` column no longer has negative values.

In [29]:
# find problematic data in the `days_employed` column and calculate the percentage
negative_values = df.loc[df['days_employed'] < 0, 'days_employed'].count()
percent_neg = negative_values/df['days_employed'].count()
print(f'The percentage of negative values for `days_employed` --> {percent_neg:.0%}\n')

# displays sample data
df.head()

The percentage of negative values for `days_employed` --> 82%



children  days_employed  dob_years            education  education_id   
0         1   -8437.673028         42    bachelor's degree             0  \
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  secondary education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income   
0            married                 0      F    employee     0     40620.102  \
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

From previous findings, there is a `negative` number in the `days_employed` column of `82%`. Considering that there are no `negative` days, it is necessary to change them to `positive` as follows:

In [30]:
# converts negative values to positive values
df['days_employed'] = df['days_employed'].abs()

In [31]:
# ensure there are no negative values anymore
negative_values = df.loc[df['days_employed'] < 0, 'days_employed'].count()
percent_neg = negative_values/df['days_employed'].count()
print(f'The percentage of negative values for `days_employed` --> {percent_neg:.0%}\n')

# displays sample data
df.head()

The percentage of negative values for `days_employed` --> 0%



children  days_employed  dob_years            education  education_id   
0         1    8437.673028         42    bachelor's degree             0  \
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income   
0            married                 0      F    employee     0     40620.102  \
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

There are no `negative` values in the `days_employed` column anymore.

In [32]:
# check `dob_years` for suspicious values

# displays head
print(df['dob_years'].value_counts().sort_index().head())

# displays tail
print(df['dob_years'].value_counts().sort_index().tail())

dob_years
0     101
19     14
20     51
21    111
22    183
Name: count, dtype: int64
dob_years
71    58
72    33
73     8
74     6
75     1
Name: count, dtype: int64


In [33]:
# persentase nilai yang bernilai nol
zero_values = df.loc[df['dob_years'] == 0]['dob_years'].count()
percent_zero  = zero_values/df['dob_years'].count()
print(f'The percentage value 0 in "dob_years" is: {percent_zero:.2%}')

The percentage value 0 in "dob_years" is: 0.47%


From the findings above, there is an odd value, namely there is still an age value of `'0'` in the `dob_years` column. Given the age, nothing is `'0'`. Next we replace it with the average value of `dob_years`.

In [34]:
# replaces the 0 values in the `dob_years` column with their average values
df.loc[df['dob_years'] == 0,'dob_years'] = int(df['dob_years'].mean())

In [35]:
# ensure there are no 0's in the `dob_years` column
zero_values = df.loc[df['dob_years'] == 0]['dob_years'].count()
percent_zero  = zero_values/df['dob_years'].count()
print(f'The percentage value 0 in "dob_years" is: {percent_zero:.2%}')

The percentage value 0 in "dob_years" is: 0.00%


In [36]:
# check `family_status` for suspicious values
df['family_status'].value_counts()

family_status
married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: count, dtype: int64

There is a value of `widow / widower` in the `family_status` column. Let's change it to a simpler value to analyze.

In [37]:
# change the value of `widow/widower` to `widowed`
df.loc[df['family_status'] == 'widow / widower','family_status'] = 'widowed'

In [38]:
# re-examine the `family_status` column which has been fixed
df['family_status'].value_counts()

family_status
married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widowed                960
Name: count, dtype: int64

the value of `widow / widower` has been changed to `widowed`.

In [39]:
# check `gender` for suspicious values
df['gender'].value_counts()

gender
F      14236
M       7288
XNA        1
Name: count, dtype: int64

From the `gender` column there are three values. To make it easier to read, we will explain again as follows:

In [40]:
# fixed the gender column according to its original meaning
df.loc[df['gender'] == 'F','gender'] = 'female'
df.loc[df['gender'] == 'M','gender'] = 'male'
df.loc[df['gender'] == 'XNA','gender'] = 'unknown'

In [41]:
# rechecking the fixed `gender` column
df['gender'].value_counts()

gender
female     14236
male        7288
unknown        1
Name: count, dtype: int64

In [42]:
# check `income_type` for suspicious values
df['income_type'].value_counts()

income_type
employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: count, dtype: int64

There is a `paternity / maternity leave` value in the `income status` column. Let's change it to a simpler value to analyze.

In [43]:
# change the value of `paternity / maternity leave` to `furlough`
df.loc[df['income_type'] == 'paternity / maternity leave','income_type'] = 'furlough'

In [44]:
# re-examine the `income_type` column which has been fixed
df['income_type'].value_counts()

income_type
employee         11119
business          5085
retiree           3856
civil servant     1459
unemployed           2
entrepreneur         2
student              1
furlough             1
Name: count, dtype: int64

the value of `paternity / maternity leave` has been changed to `furlough`.

**Checking Duplicate Data**

Next, let's check whether there are duplicates in the dataset that we have.

In [45]:
# check for duplicate datasets
df.duplicated().sum()

71

There are `71` duplicates in the `original dataframe`. We must remove these duplicates to get accurate results because they will affect the calculation.

In [46]:
# resolve duplicate values
df.drop_duplicates(inplace=True)

In [47]:
# recheck for the missing values
df.duplicated().sum()

0

# Work with missing values <a id= 'cont_5' ></a>

Dictionary is used to facilitate the reading of a data or value. Here we want to load the dictionary found in the dataset. The dictionaries that we will create are `education` and `family_status`.

In [48]:
# dictionary of education
edu_dict = dict(zip(df.education_id,df.education))
edu_dict

{0: "bachelor's degree",
 1: 'secondary education',
 2: 'some college',
 3: 'primary education',
 4: 'graduate degree'}

In [49]:
# dictionary of family_status
fam_dict = dict(zip(df.family_status_id,df.family_status))
fam_dict

{0: 'married',
 1: 'civil partnership',
 2: 'widowed',
 3: 'divorced',
 4: 'unmarried'}

### Fixed missing value in `total income` <a id= 'cont_6' ></a>

To fix the `missing` value in the `total_income` column, the following strategies are needed:
- Identify which is effective for filling in missing values whether with `mean` or `median`
- Grouping customer categories based on their age `dob_years` in a new column
- Calculates the `mean` and `median` from the `total_income` column for the customer group category
- Determines whether `mean` or `median` is used
- Fill in `missing` values
- Checks if there are still `missing` values

In [50]:
# create a function to categorize age
def age_category(age):
    if isinstance(age, (int, float, complex)):
        if age < 25:
            return 'teenager'
        elif 25 <= age < 60:
            return 'adult'
        elif age >= 60:
            return 'old'
    else:
        return 'unknown'

In [51]:
# test the function that has been created to categorize age

# exam for teenager
budi = 18
print(f'category : {age_category(budi)}')

# exam for adult
dudung = 34
print(f'category : {age_category(dudung)}')

# exam for old
andi = 65
print(f'category : {age_category(andi)}')

category : teenager
category : adult
category : old


In [52]:
# create a new column based on age category
df['age_category'] = df['dob_years'].agg(age_category)

In [53]:
# checks the new column that has been created
df['age_category'].value_counts()

age_category
adult       18079
old          2500
teenager      875
Name: count, dtype: int64

Many factors affect the income, but this time we will discuss several factors that affect `total_income` to determine which `mean` or `median` value to use.

In [54]:
# filter the dataset without missing values
df_without_nan = df.dropna(subset=['days_employed','total_income'])
print(f'{df_without_nan.isna().sum()}\n')

# returns a filtered sample dataset
df_without_nan.head()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_category        0
dtype: int64



children  days_employed  dob_years            education  education_id   
0         1    8437.673028         42    bachelor's degree             0  \
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id  gender income_type  debt   
0            married                 0  female    employee     0  \
1            married                 0  female    employee     0   
2            married                 0    male    employee     0   
3            married                 0    male    employee     0   
4  civil partnership                 1  female     retiree     0   

   total_income                  purpose age_category  
0     40620.102    purchase of the house        adult  
1     17932.802             car purchase        adult  
2     23341.752    purchase of the house        adult  
3     42820.568  supplementary education        adult  
4     25378.572        to have a wedding        adult

In [55]:
# calculates the average value for income by age category factor
mean_income = df_without_nan.groupby('age_category')['total_income'].mean()
mean_income

age_category
adult       27503.004629
old         23021.639994
teenager    22703.351103
Name: total_income, dtype: float64

In [56]:
# calculates the median value for income by age category factor
median_income = df_without_nan.groupby('age_category')['total_income'].median()
median_income

age_category
adult       23857.962
old         19761.425
teenager    20572.209
Name: total_income, dtype: float64

In [57]:
# calculates the average value for income by education
mean_income_edu = df_without_nan.groupby('education')['total_income'].mean()
mean_income_edu

education
bachelor's degree      33142.802434
graduate degree        27960.024667
primary education      21144.882211
secondary education    24594.503037
some college           29045.443644
Name: total_income, dtype: float64

In [58]:
# calculates the median value for income by education
median_income_edu = df_without_nan.groupby('education')['total_income'].median()
median_income_edu

education
bachelor's degree      28054.5310
graduate degree        25161.5835
primary education      18741.9760
secondary education    21836.5830
some college           25618.4640
Name: total_income, dtype: float64

In [59]:
# calculates the average value for income by family_status
mean_income_fam = df_without_nan.groupby('family_status')['total_income'].mean()
mean_income_fam

family_status
civil partnership    26694.428597
divorced             27189.354550
married              27041.784689
unmarried            26934.069805
widowed              22984.208556
Name: total_income, dtype: float64

In [60]:
# calculates the median value for income by family_status
median_income_fam = df_without_nan.groupby('family_status')['total_income'].median()
median_income_fam

family_status
civil partnership    23186.534
divorced             23515.096
married              23389.540
unmarried            23149.028
widowed              20514.190
Name: total_income, dtype: float64

From the several tests that have been carried out, it can be concluded that the `mean` and `median` values of `total_income` based on the several factors tested have values that are not much different. This shows that the values are almost uniform. So the `mean` will be applied to fill in the missing values.

In [61]:
# create a function to fill in missing values
def fill_na_category(data,col,value):
    data[col] = data[col].fillna(value)
    return data[col]

In [62]:
# calculates the average value in the `total income` column
mean_all = df_without_nan['total_income'].mean()
mean_all

26787.568354658673

In [63]:
# apply the function to each row
fill_na_category(df,'total_income',mean_all)

0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21520    35966.698
21521    24959.969
21522    14347.610
21523    39054.888
21524    13127.587
Name: total_income, Length: 21454, dtype: float64

In [64]:
# checks if the missing values are still in the `total income` column
df['total_income'].isna().sum()

0

After fixing the `missing` values in the `total_income` column, the total values in the `total_income` column are the same as the sum of the other columns except for the `days_employed` column which will be corrected afterwards.

###  Fixed the value in `days_employed` <a id= 'cont_7' ></a>

In [65]:
# calculates the median of `days_employed`
median_days = df_without_nan.groupby('age_category')['days_employed'].median()
median_days

age_category
adult         1978.781961
old         355229.618218
teenager       744.542130
Name: days_employed, dtype: float64

In [66]:
# calculates the average of `days_employed`
mean_days = df_without_nan.groupby('age_category')['days_employed'].mean()
mean_days

age_category
adult        39695.452041
old         286544.143436
teenager      1282.650421
Name: days_employed, dtype: float64

From the results of calculations to compare the `mean` and `median` values based on the customer's age category, it can be seen that the values in the `days_employed` column are not uniform in each category. This proves that the values are not concentrated to one center so that the `mean` value cannot be used to fill in missing values. The `median` value that can fill in the missing values.

In [67]:
# create a function to get the median value
def median_days(data,col):
    calc_median = data[col].median(skipna=True)
    return calc_median

In [68]:
# try the function does it work?
row = [45,np.nan,66,np.nan,77,88]
col = ['coba']
try_df = pd.DataFrame(data=row,columns=col)
median_days(try_df,'coba')

71.5

The function is already working.

In [69]:
# apply a function to days_employed to get the median value
calc_median_days = median_days(df,'days_employed')

In [70]:
# replaces missing values in the `days_employed` column with their median values
fill_na_category(df,'days_employed',calc_median_days)

0          8437.673028
1          4024.803754
2          5623.422610
3          4124.747207
4        340266.072047
             ...      
21520      4529.316663
21521    343937.404131
21522      2113.346888
21523      3112.481705
21524      1984.507589
Name: days_employed, Length: 21454, dtype: float64

In [71]:
# checks if the missing values are still present in the `days_employed` column
df['days_employed'].isna().sum()

0

For ease of reading we will change the data type in the `days_employed` and `total_income` columns to `integer` as follows:

In [72]:
# change the category data type from float to integer
df['days_employed'] = df['days_employed'].astype('int')
df['total_income'] = df['total_income'].astype('int')

In [73]:
# examine the entire dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21454 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21454 non-null  int64 
 1   days_employed     21454 non-null  int32 
 2   dob_years         21454 non-null  int64 
 3   education         21454 non-null  object
 4   education_id      21454 non-null  int64 
 5   family_status     21454 non-null  object
 6   family_status_id  21454 non-null  int64 
 7   gender            21454 non-null  object
 8   income_type       21454 non-null  object
 9   debt              21454 non-null  int64 
 10  total_income      21454 non-null  int32 
 11  purpose           21454 non-null  object
 12  age_category      21454 non-null  object
dtypes: int32(2), int64(5), object(6)
memory usage: 2.1+ MB


After fixing the `missing` values in the `days_employed` column, the total values in the `days_employed` column are the same as the sum of the other columns.

## Data Categorization <a id= 'cont_8' ></a>

Next, let's create some additional columns in the form of categorizing data using `Category Data` and `Numerical Data` as follows:

**Category Data**

In [74]:
# selects the column for which the category will be created
df_categories = df[['children','debt','purpose']]

# displays sample data
df_categories.head()

children  debt                  purpose
0         1     0    purchase of the house
1         1     0             car purchase
2         0     0    purchase of the house
3         3     0  supplementary education
4         0     0        to have a wedding

In [75]:
# checks the unique value of `children`
df['children'].unique()

array([ 1,  0,  3,  2,  4, 20,  5], dtype=int64)

In [76]:
# checks the unique value of `debt`
df['debt'].unique()

array([0, 1], dtype=int64)

In [77]:
# checks the unique value of `purpose`
df['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

There are several columns that we will recategorize for easier reading and we will create new columns based on these categorizations. We categorize the `children` and `debt` columns based on the dictionary, while the `purpose` column we categorize based on generality.

In [78]:
# create functions to categorize data based on common topics

# For the category of children_status
def children(children):
    if children == 0:
        return 'have no children' 
    else:
        return 'have children'

# For the category of  debt_status
def debt_status(debt):
    if debt == 0:
        return 'on time'
    else:
        return 'default'

# For the category of purpose_status
def purpose_status(purpose):
    if 'cars' in purpose:
        return 'car purpose'
    elif 'car' in purpose:
        return 'car purpose'
    elif 'education' in purpose:
        return 'educational purpose'
    elif 'university' in purpose:
        return 'educational purpose'
    elif 'educated' in purpose:
        return 'educational purpose'
    elif 'wedding' in purpose:
        return 'wedding purposes'
    elif 'house' in purpose:
        return 'property purpose'
    elif 'housing' in purpose:
        return 'property purpose'
    elif 'estate' in purpose:
        return 'property purpose'
    elif 'property' in purpose:
        return 'property purpose'

In [79]:
# apply each function to create a new column
df['children_status'] = df['children'].agg(children)
df['debt_status'] = df['debt'].agg(debt_status)
df['purpose_status'] = df['purpose'].agg(purpose_status)

In [80]:
# displays the results of the `children status` column
df['children_status'].value_counts()

children_status
have no children    14091
have children        7363
Name: count, dtype: int64

In [81]:
# displays the results of the `debt_status` column 
df['debt_status'].value_counts()

debt_status
on time    19713
default     1741
Name: count, dtype: int64

In [82]:
# displays the results of the `purpose_status` column
df['purpose_status'].value_counts()

purpose_status
property purpose       10811
car purpose             4306
educational purpose     4013
wedding purposes        2324
Name: count, dtype: int64

**Numerical Data**

In [83]:
# select column `total_income` as numeric data
df_numeric = df[['total_income']]

# displays sample data
df_numeric.head()

total_income
0         40620
1         17932
2         23341
3         42820
4         25378

In [84]:
# get the mean and median values in the `total income` column

dict_income = {'total_income':['mean','median']}
stat_income_new = df.agg(dict_income)

mean_income_new = int(stat_income_new['total_income']['mean'])
median_income_new = int(stat_income_new['total_income']['median'])

print(f"Mean income value : {mean_income_new}")
print(f"Median income value : {median_income_new}")

Mean income value : 26787
Median income value : 24966


For the numeric category we will group `total_income` based on the `mean` value of `total_income`. the `mean` value was chosen because the data or `total_income` value is concentrated towards the same center, namely the `mean`.

In [85]:
# function to categorize numeric groups into categories
def income_status(income):
    if income < mean_income_new:
        return 'below average income'
    else:
        return 'above average income'

In [86]:
# create a new categorical column from the `total income` column
df['income_status'] = df['total_income'].agg(income_status)

In [87]:
# displays the results of the `total income` column
df['income_status'].value_counts()

income_status
below average income    11892
above average income     9562
Name: count, dtype: int64

In [88]:
# displays the final results of data categorization
df.head()

children  days_employed  dob_years            education  education_id   
0         1           8437         42    bachelor's degree             0  \
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id  gender income_type  debt   
0            married                 0  female    employee     0  \
1            married                 0  female    employee     0   
2            married                 0    male    employee     0   
3            married                 0    male    employee     0   
4  civil partnership                 1  female     retiree     0   

   total_income                  purpose age_category   children_status   
0         40620    purchase of the house        adult     have children  \
1         17932             car purchase        adult     have children   
2         23341    purchase of the house        adult  have no children   
3         42820  supplementary education        adult     have children   
4         25378        to have a wedding        adult  have no children   

  debt_status       purpose_status         income_status  
0     on time     property purpose  above average income  
1     on time          car purpose  below average income  
2     on time     property purpose  below average income  
3     on time  educational purpose  above average income  
4     on time     wedding purposes  below average income

## Checking the hypothesis <a id= 'cont_9' ></a>


Finally, let's test some of the hypotheses that have been made before using the data that we have prepared beforehand.

**Hypothesis 1: Is there a correlation between having children and the probability of defaulting?** <a id= 'cont_10' ></a>

In [89]:
# check child data and default data
debt_status_children = pd.DataFrame(df.groupby('children_status')['debt_status'].value_counts())
debt_status_children.reset_index(inplace=True)
debt_status_children

children_status debt_status  count
0     have children     on time   6685
1     have children     default    678
2  have no children     on time  13028
3  have no children     default   1063

In [90]:
# calculates the default percentage based on the number of children
group_count = debt_status_children.groupby('children_status')['count'].sum()
percent_chld = debt_status_children[debt_status_children['debt_status']=='default'].groupby('children_status')['count'].sum()
percent_chld = round(percent_chld/group_count*100,2)
percent_chld.map('{:.2f} %'.format)

children_status
have children       9.21 %
have no children    7.54 %
Name: count, dtype: object

We can conclude that the percentage of `Default` for customers who `Have Children` is higher compared to customers who `Do Not Have Children` with a value of `9.21%` and `7.54%` respectively.

**Hypothesis 2: Is there a correlation between family status and the probability of default?** <a id= 'cont_11' ></a>

In [91]:
# check family status data and default data
debt_status_family = pd.DataFrame(df.groupby('family_status')['debt_status'].value_counts())
debt_status_family.reset_index(inplace=True)
debt_status_family

family_status debt_status  count
0  civil partnership     on time   3763
1  civil partnership     default    388
2           divorced     on time   1110
3           divorced     default     85
4            married     on time  11408
5            married     default    931
6          unmarried     on time   2536
7          unmarried     default    274
8            widowed     on time    896
9            widowed     default     63

In [92]:
# calculate the default percentage based on family status
group_count = debt_status_family.groupby('family_status')['count'].sum()
percent_fml = debt_status_family[debt_status_family['debt_status']=='default'].groupby('family_status')['count'].sum()
percent_fml = round(percent_fml/group_count*100,2)
percent_fml.map('{:.2f} %'.format)

family_status
civil partnership    9.35 %
divorced             7.11 %
married              7.55 %
unmarried            9.75 %
widowed              6.57 %
Name: count, dtype: object

We can conclude that the highest to lowest percentage of `Default to Pay` for customers based on `Family Status` is `unmarried` status of `9.75%`, `civil partnership` status - `9.35%`, `married` - `7.55% ` status `divorced` - `7.11 %` and `widowed` status - `6.57 %`.

**Hypothesis 3: Is there a correlation between the level of income and the probability of default?** <a id= 'cont_12' ></a>

In [93]:
# calculate income level data and default data
debt_status_income = pd.DataFrame(df.groupby('income_status')['debt_status'].value_counts())
debt_status_income.reset_index(inplace=True)
debt_status_income

income_status debt_status  count
0  above average income     on time   8824
1  above average income     default    738
2  below average income     on time  10889
3  below average income     default   1003

In [94]:
# calculate default percentage based on income level
group_count = debt_status_income.groupby('income_status')['count'].sum()
percent_inc = debt_status_income[debt_status_income['debt_status']=='default'].groupby('income_status')['count'].sum()
percent_inc = round(percent_inc/group_count*100,2)
percent_inc.map('{:.2f} %'.format)

income_status
above average income    7.72 %
below average income    8.43 %
Name: count, dtype: object

We can conclude that the percentage of `Default to Pay` for customers who have `Below Average Income` is higher than for customers who have `Above Average Income` with respective values of `8.43%` and `7.72%`.

**Hypothesis 4: How do installment goals affect the default rate?** <a id= 'cont_13' ></a>

In [95]:
# calculate credit destination data and default data
debt_status_purpose = pd.DataFrame(df.groupby('purpose_status')['debt_status'].value_counts())
debt_status_purpose.reset_index(inplace=True)
debt_status_purpose

purpose_status debt_status  count
0          car purpose     on time   3903
1          car purpose     default    403
2  educational purpose     on time   3643
3  educational purpose     default    370
4     property purpose     on time  10029
5     property purpose     default    782
6     wedding purposes     on time   2138
7     wedding purposes     default    186

In [96]:
# calculate the percentage of default rate for each installment goal and do the analysis
group_count = debt_status_purpose.groupby('purpose_status')['count'].sum()
percent_prps = debt_status_purpose[debt_status_purpose['debt_status']=='default'].groupby('purpose_status')['count'].sum()
percent_prps = round(percent_prps/group_count*100,2)
percent_prps.map('{:.2f} %'.format)

purpose_status
car purpose            9.36 %
educational purpose    9.22 %
property purpose       7.23 %
wedding purposes       8.00 %
Name: count, dtype: object

We can conclude that the highest to lowest percentage of `Default` for customers based on `Credit Purpose` is `car needs` - `9.36%`, `educational needs` - `9.22%`, `marriage needs` -` 8.00 %` and `property purpose` - `7.23%`.

# General conclusion <a id= 'cont_14' ></a>

`Default to pay` is a phenomenon where the customer is unable to pay on time, so analysis is needed to find out which category causes `Default to pay` to occur. Analysis requires relevant data related to the `Default to Pay` phenomenon starting from the process of identifying the number of rows and columns in the dataframe. Correct the value in the category whether it is correct or needs to be corrected. Then a search for missing values is carried out in the dataframe, a search for reasons why missing values can occur. Is the missing value a random event or a pattern? Check for patterns related to missing values, if indeed the missing values are not random then look for other values based on statistics such as the `mean` and `median` of the `total_income` and `days_employed` columns respectively. Then after obtaining a replacement value, input the value to the missing value.

After the repair process is carried out, an analysis process is carried out to test the hypotheses that have been made. The first step is to categorize the data. Searching for unique values and creating new categories can make analysis easier. New columns are created here, namely `debt_status`, `children_status`, and `purpose_status` to make it easier to group data. Next is to create a dictionary based on data that has an ID and a description of that ID. Here the dictionary created is the `education` dictionary and the `family_status` dictionary.

The final stage is testing the hypothesis including the following:
- The relationship between `Default to Pay` and `Number of Children` shows that customers who `Have Children` have a higher score compared to customers who `No Children` which is `9.21 %`.
- The relationship between `Default to Pay` and `Family Status` shows that the highest customer is `unmarried` at `9.75%`.
- The relationship between `Default to Pay` and `Income Level` shows that the highest customer is a customer with `Below Average Income Level` which is equal to `8.43%`.
- The relationship between `Default to Pay` and `Installment Purpose` is obtained that the highest customer is for `Car Needs` which is equal to `9.36%`.